mqtt/mqtt-broker/src/auth.rs

```diff
@@ -0,0 +1,71 @@
+use derive_more::Display;
+use failure::Fail;
+
+use crate::Error;
+
+#[derive(Clone, Debug, Display, PartialEq)]
+pub enum AuthId {
+    #[display(fmt = "*")]
+    Any,
+
+    Value(Identity),
+}
+
+pub type Identity = String;
+
+#[derive(Clone, Debug)]
+pub struct Certificate(Vec<u8>);
+
+#[allow(dead_code)]
+pub struct Credentials {
+    username: Option<String>,
+    password: Option<String>,
+    certificate: Option<Certificate>,
```

Comment:
Do you suggest I split them? Like this
```rust
pub enum Credentials { Basic(String, String), Certificate(Certificate) }
```

Reply:
If that makes sense, then yeah. Basically I would move all of the validation to the constructor of this type, so that the rest of the code can just assume everything is nice.